Mathlib/NumberTheory/NumberField/Completion.lean

```diff
@@ -0,0 +1,344 @@
+/-
+Copyright (c) 2024 Salvatore Mercuri. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Salvatore Mercuri
+-/
+import Mathlib.Algebra.Field.Subfield
+import Mathlib.Analysis.Normed.Module.Completion
+import Mathlib.NumberTheory.NumberField.Embeddings
+import Mathlib.Topology.UniformSpace.Basic
+import Mathlib.Topology.Instances.Real
+
+/-!
+# The completion of a number field at an infinite place
+
+This file contains the completion of a number field at an infinite place.
+
+## Main definitions
+ - `WithAbs` : type synonym equipping a semiring with an absolute value.
+ - `AbsoluteValue.completion` : the uniform space completion of a field `K` equipped with real
+  absolute value.
+ - `NumberField.InfinitePlace.completion` : the completion of a number field `K` at an infinite
+  place, obtained by completing `K` with respect to the absolute value associated to the infinite
+  place.
+ - `NumberField.InfinitePlace.Completion.extensionEmbedding` : the embedding `v.embedding : K →+* ℂ`
+  extended to `v.completion →+* ℂ`.
+ - `NumberField.InfinitePlace.Completion.extensionEmbedding_of_isReal` : if the infinite place `v`
+  is real, then this extends the embedding `v.embedding_of_isReal : K →+* ℝ` to
+  `v.completion →+* ℝ`.
+ - `NumberField.InfinitePlace.Completion.equiv_real_of_isReal` : the ring isomorphism
+  `v.completion ≃+* ℝ` when `v` is a real infinite place; the forward direction of this is
+  `extensionEmbedding_of_isReal`.
+ - `NumberField.InfinitePlace.Completion.equiv_complex_of_isComplex` : the ring isomorphism
+  `v.completion ≃+* ℂ` when `v` is a complex infinite place; the forward direction of this is
+  `extensionEmbedding`.
+
+## Main results
+ - `NumberField.Completion.locallyCompactSpace` : the completion of a number field at
+  an infinite place is locally compact.
+ - `NumberField.Completion.isometry_extensionEmbedding` : the embedding `v.completion →+* ℂ` is
+  an isometry. See also `isometry_extensionEmbedding_of_isReal` for the corresponding result on
+  `v.completion →+* ℝ` when `v` is real.
+ - `NumberField.Completion.bijective_extensionEmbedding_of_isComplex` : the embedding
+  `v.completion →+* ℂ` is bijective when `v` is complex. See also
+  `bijective_extensionEmebdding_of_isReal` for the corresponding result for `v.completion →+* ℝ`
+  when `v` is real.
+
+## Tags
+number field, embeddings, infinite places, completion
+-/
+noncomputable section
+
+/-- Type synonym equipping a semiring with an absolute value. -/
+@[nolint unusedArguments]
+def WithAbs {R S} [Semiring R] [OrderedSemiring S] : AbsoluteValue R S → Type _ := fun _ => R
+
+namespace WithAbs
+
+instance {R} [Ring R] (v : AbsoluteValue R ℝ) : NormedRing (WithAbs v) where
```

Comment:
Why not making this a `def`, without using `WithAbs`? (It should be in some other file I think).

Reply:
Do you mean doing something like the following? 

```
def normedRing {R : Type*} [Ring R] (v : AbsoluteValue R ℝ) : NormedRing R where 
  dist_eq _ _ := rfl
  dist_self x := by simp only [sub_self, MulHom.toFun_eq_coe, AbsoluteValue.coe_toMulHom, map_zero]
  dist_comm := v.map_sub
  dist_triangle := v.sub_le
  norm_mul x y := (v.map_mul x y).le
  eq_of_dist_eq_zero := by simp only [MulHom.toFun_eq_coe, AbsoluteValue.coe_toMulHom,
    AbsoluteValue.map_sub_eq_zero_iff, imp_self, implies_true]

instance {R : Type*} [Ring R] (v : AbsoluteValue R ℝ) : NormedRing (WithAbs v) := 
  letI : Ring (WithAbs v) := inferInstanceAs (Ring R)
  normedRing v
```

and placing the `normedRing` `def` in an `AbsoluteValue` related file?

Reply:
Yes, you can call this something like `AbsoluteValue.normedRing` so people can write `v.normedRing`. But it is just an idea, I am not sure it will make things easier.

Reply:
It seems to me that the situation is analogous to [Valuation](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Valuation/Basic.html#Valuation)[Valuation](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Valuation/Basic.html#Valuation) and [Valued](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Algebra/Valued/ValuationTopology.html#Valued)[Valued](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Algebra/Valued/ValuationTopology.html#Valued). It seems that mathlib's strategy is to use `Valuation.mk` to turn a valuation into an instance of `Valued` (something you can do locally in a file) and then play with it. I think that here the situation is the same, with a valuation replaced by an absolute value and `Valued` replaced by `NormedRing`.

Reply:
Yes, this situation is analogous to the non-Archimedean case. Actually, in my original approach in #13577  I used the model of [`Valued`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Algebra/Valued/ValuationTopology.html#Valued) and [`HeightOneSpectrum.adicCompletion`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/DedekindDomain/AdicValuation.html#IsDedekindDomain.HeightOneSpectrum.adicCompletion). This worked fine, but then I needed to use either lots of `@`s or `letI := v.normedField`s everywhere to explicitly tell the system which uniform space structure I needed. The type synonym approach of `WithAbs` lets the inference system take care of this, so it seems a bit neater to me, but we can keep `WithAbs` and its instances while abstracting the `normedRing` definitions to an `AbsoluteValue` file/namespace as you suggest. I will take a look!

Reply:
Oh, this is strange and it might showcase some deficiency in the `Valued` API. Have you tried there to define a `local instance` of uniform space using `v.normedField`? I do have the feeling that having both valuations and absolute values (that are the same gadget, after all) is a problem for the library. Or do you find any advantage of using the latter rather than the former (beyond the problem that `InfinitePlaces` are defined using `AbsoluteValue` but this can be refactored)?

Reply:
Yes, that's true.